[SPARK-8171] [Web UI] Javascript based infinite scrolling for the log page

ajbozarth opened the pull request:
Updated the log page by replacing the current pagination with a javascript-based infinite scroll solution

ajbozarth committed:
simulated infinite scrolling using current pagination code

ajbozarth commented:
I spent some time looking into ways to use javascript to load the content rather than use a page refresh and all the solutions I found required that the file has to be available on the web-server itself, which is very difficult to do with a continuously updating log file.

Instead I implemented a alteration to the current pagination that adds to the top (or bottom) of the log a page at a time rather than limiting you to read only a page at a time. By setting the scroll starting location at page load this simulates an infinite scroll even though the page is re-loading.

I also set the log to scroll to the bottom on initial page load.

Given this is inspired by but not exactly the original jira and does distinctly change the use of this page I would love to get peoples opinions on it's usability, look and feel, and if it's worth the change at all.

JoshRosen commented:
Screenshot? Please include them for all UI changes.

ajbozarth commented:
Sorry took me a bit longer to get them ready than usual.

Initial page load:
![initial load](https://cloud.githubusercontent.com/assets/13952758/12569727/454a544a-c386-11e5-95aa-5b42da9998c7.png)

initial load scrolled to top:
![initial top](https://cloud.githubusercontent.com/assets/13952758/12569732/52cfb894-c386-11e5-81d2-f9fcd0e70e53.png)

after "load more" is clicked and page reloads:
![after load more](https://cloud.githubusercontent.com/assets/13952758/12569739/5ff8b2dc-c386-11e5-90af-92c3bf11a112.png)

if the full file is loaded:
![top of file](https://cloud.githubusercontent.com/assets/13952758/12569746/6d94e046-c386-11e5-9ba4-773c62ec50ca.png)

when you scroll to the bottom after loading more and there is new log available:
![load new](https://cloud.githubusercontent.com/assets/13952758/12569759/83fa63ce-c386-11e5-976e-33073d4c1de9.png)

SparkQA commented:
**[Test build #50050 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50050/consoleFull)** for PR 10910 at commit [`4d7c433`](https://github.com/apache/spark/commit/4d7c43373d43126b488540d7659274277665f51c).

SparkQA commented:
**[Test build #50050 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50050/consoleFull)** for PR 10910 at commit [`4d7c433`](https://github.com/apache/spark/commit/4d7c43373d43126b488540d7659274277665f51c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50050/
Test FAILed.

ajbozarth commented:
Retest this please

SparkQA commented:
**[Test build #50065 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50065/consoleFull)** for PR 10910 at commit [`4d7c433`](https://github.com/apache/spark/commit/4d7c43373d43126b488540d7659274277665f51c).

SparkQA commented:
**[Test build #50065 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50065/consoleFull)** for PR 10910 at commit [`4d7c433`](https://github.com/apache/spark/commit/4d7c43373d43126b488540d7659274277665f51c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50065/
Test PASSed.

holdenk commented:
So every time the user clicks "load more" it will request/fetch all of the log that has already been loaded again? (Just trying to understand the design).

ajbozarth commented:
When the user clicks load more the page refreshes and on the new page load it reads a larger chunk of the file.
The main implementation difference between the previous version and this is that when the user requests another page of the log it is added to the current log page(s) rather than just showing the new requested page alone.

JoshRosen commented:
This isn't quite what I had in mind when filing that JIRA; the UX that I envisioned was something where if a user clicks "load more" the Javascript requests only the new / additional portion log output from the backend then update the UI without reloading the page.

ajbozarth commented:
I had that in mind too, I just had this implementation in my head as an alternative and wanted to see what you thought. If you'd rather stick with what we have for now and wait until we figure out a js solution, then we can close this PR. I'll also spend just a bit more time trying to find a js solution to fetching the file data before moving to another task.

JoshRosen commented:
Yeah, let's close this for the time being and re-open when you have a new approach which is ready to be reviewed.

ajbozarth committed:
Merge branch 'master' into spark8171

ajbozarth committed:
updated log page to use javascript for infinite scrolling

ajbozarth committed:
Merge branch 'master' into spark8171

ajbozarth commented:
Thanks @JoshRosen for rejecting my first version, making it did helped me find this new solution using js  though. 

Few questions: 
Should we load the content of the log-data span and/or the logText content dynamically on initial page load or leave it hard-coded as is?
Would we rather have a proper infinite scroll where it detects when the user reaches the top or bottom rather than the buttons I've implemented? I prefer the buttons myself.

ajbozarth commented:
I'll post some new screen shots below but the look and feel is basically the same as the previous fake version. 
I've changed the button text a bit and added an transient info alert when the user tries to load new and nothing is available. Note it doesn't replace the button but appears above it.

![initialloadclick](https://cloud.githubusercontent.com/assets/13952758/12694354/f5cf8d9c-c6df-11e5-9478-d7fafde181b2.png)
![bottomoflog](https://cloud.githubusercontent.com/assets/13952758/12694356/f5f17dda-c6df-11e5-8a9c-e7108450cff9.png)
![endoflog](https://cloud.githubusercontent.com/assets/13952758/12694355/f5f0ffb8-c6df-11e5-9e85-60ea5f302de3.png)

ajbozarth commented:
@ckadner this was what I was showing you if you want to take a look

SparkQA commented:
**[Test build #50439 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50439/consoleFull)** for PR 10910 at commit [`35e08c7`](https://github.com/apache/spark/commit/35e08c7d3f7b89a04405795aa806cf5bbf76d9ec).

SparkQA commented:
**[Test build #50439 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50439/consoleFull)** for PR 10910 at commit [`35e08c7`](https://github.com/apache/spark/commit/35e08c7d3f7b89a04405795aa806cf5bbf76d9ec).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50439/
Test PASSed.

andrewor14 commented:
@sarutak

ajbozarth commented:
@JoshRosen opinions on the updated version?

ajbozarth commented:
Just in case it fell through the cracks last week, what do people think of the new javascript-based log page?

ajbozarth commented:
@tgravescs thanks for all the help on those other PRs, mind taking a look at this one?

ajbozarth commented:
@sarutak @JoshRosen @tgravescs in case this slipped past you guys due to Spark Summit last week

ajbozarth commented:
@JoshRosen If you have time to take a second look at this, it's been sitting in limbo for a while, thanks

ajbozarth commented:
@sarutak @JoshRosen @tgravescs I know you guys are busy getting ready for Spark Summit West, but I was wondering if you had an eta on looking at this? It's been open for a couple months now, thanks.

sarutak commented:
@ajbozarth I'll check it soon.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+
+var byteLength
+var btnHeight = 30
+
```
It's better to append `;` to all the variable declarations above.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+
+function tailLog() {
+  $(".log-content").scrollTop($(".log-content")[0].scrollHeight);
+}
```
It's not good to repeat a query for the same dom-objects like `$(.log-content)`.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+}
+
+function disableMoreButton() {
+  $(".log-more-btn").attr("disabled", "disabled");;
```
`;` is redundant.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+        cleanData = cleanData.substring(0, startByte);
+        disableMoreButton();
+      }
+      $("pre", ".log-content").prepend(cleanData);
```
I think we should prepend `cleanData + \n`.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+            var retLogLength = dataInfo[2];
+
+            var cleanData = data.substring(data.indexOf('\n') + 1).trim();
+            $("pre", ".log-content").append(cleanData);
```
Need to append `'\n' + cleanData`.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+  $.ajax({
+    type: "GET",
+    url: "/log" + baseParams,
+    success: function (data) {
```
All we need to do here is to check whether the log size is bigger. So, can we somehow check it without receiving the first 100 KB of the log?

ajbozarth replied:
I've added + "&byteLength=0" so the only data returned will be the info line, or we're you thinking something different?

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+            var retEndByte = dataInfo[1];
+            var retLogLength = dataInfo[2];
+
+            var cleanData = data.substring(data.indexOf('\n') + 1).trim();
```
I think we should not trim.

ajbozarth replied:
I was able to remove the ```trim```s by changing that line in JettyUtils, I'm currently looking into and testing it, but wouldn't this change the expected results of other HTTP requests to the API?

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+      var retStartByte = dataInfo[0];
+      var retLogLength = dataInfo[2];
+
+      var cleanData = data.substring(data.indexOf('\n') + 1).trim();
```
We don't need trim.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+          type: "GET",
+          url: "/log" + baseParams + "&byteLength=" + newDataLen,
+          success: function (data) {
+            var dataInfo = data.substring(0, data.indexOf('\n')).match(/\d+/g);
```
It's better not to repeat `data.substring` and `data.indexOf` in this `success` function.
I think we can replace this logic with `split` instead of `substring` and `indexOf`.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+    url: "/log" + baseParams + "&offset=" + offset + "&byteLength=" + byteLength,
+    success: function (data) {
+      var oldHeight = $(".log-content")[0].scrollHeight;
+      var dataInfo = data.substring(0, data.indexOf('\n')).match(/\d+/g);
```
As I mentioned below, we should not repeat `substring` and `indexOf`.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+
+      var cleanData = data.substring(data.indexOf('\n') + 1).trim();
+      if (retStartByte == 0) {
+        cleanData = cleanData.substring(0, startByte);
```
This line is not needed.

ajbozarth replied:
This did what the Math.min you suggest above did, but after getting the data (much cleaner with the Math.min, thanks)

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+
+function loadMore() {
+  var offset = Math.max(startByte - byteLength, 0);
+  var newLogLength = Math.min(curLogLength + byteLength, totalLogLength);
```
`newLogLength` is not used.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,125 @@
+
+  $.ajax({
+    type: "GET",
+    url: "/log" + baseParams + "&offset=" + offset + "&byteLength=" + byteLength,
```
We should receive `Math.min(byteLength, startByte)` bytes of data instead of `byteLength` bytes.

ajbozarth replied:
This is much cleaner, thanks

sarutak commented:
@ajbozarth Sorry for having you wait. I've inspected quickly and left some comments.

@JoshRosen I think this feature itself is useful but is this the feature  you had in mind?

ajbozarth commented:
@sarutak thanks for the feedback, I'll look through and address them as I have time over the next couple days

ajbozarth committed:
Merge branch 'master' into spark8171

ajbozarth commented:
@sarutak I just went through your comments and responded and made fixes. I'm currently testing locally and will push my changes later today.

ajbozarth committed:
Addressed js comments

SparkQA commented:
**[Test build #55251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55251/consoleFull)** for PR 10910 at commit [`0f49738`](https://github.com/apache/spark/commit/0f49738de032c78873063f29cf60c060ab3b04ef).

SparkQA commented:
**[Test build #55251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55251/consoleFull)** for PR 10910 at commit [`0f49738`](https://github.com/apache/spark/commit/0f49738de032c78873063f29cf60c060ab3b04ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55251/
Test PASSed.

ajbozarth commented:
@sarutak @JoshRosen What do you think of the latest changes?

sarutak commented:
@ajbozarth Thank you for the update. I'll check it next Sunday or Monday.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,129 @@
+    url: "/log" + baseParams + "&offset=" + offset + "&byteLength=" + moreByteLength,
+    success: function (data) {
+      var oldHeight = $(".log-content")[0].scrollHeight;
+      var newlineIndex = data.indexOf('\n')
```
nit: semicolon.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,129 @@
+}
+
+function noNewAlert() {
+  var alert = $(".no-new-alert")
```
nit: semicolon.

sarutak commented on core/src/main/resources/org/apache/spark/ui/static/log-view.js:
```diff
@@ -0,0 +1,129 @@
+          type: "GET",
+          url: "/log" + baseParams + "&byteLength=" + newDataLen,
+          success: function (data) {
+            var newlineIndex = data.indexOf('\n')
```
nit: semicolon.

sarutak commented:
@ajbozarth I've left some comments about very trivial nits, otherwise LGTM.

sarutak commented on core/src/main/scala/org/apache/spark/deploy/worker/ui/LogPage.scala:
```diff
@@ -76,49 +74,44 @@ private[ui] class LogPage(parent: WorkerWebUI) extends WebUIPage("logPage") with
-          Previous 0 B
-        </button>
-      }
+    val curLogLength = endByte - startByte;
```
nit:semicolon is not needed here.

ajbozarth committed:
fixed some semicolons

ajbozarth commented:
Fixed the semicolons, switching back and forth between js and scala when they're "optional" on both really messes with you. Thanks @sarutak

SparkQA commented:
**[Test build #56098 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56098/consoleFull)** for PR 10910 at commit [`293f1a4`](https://github.com/apache/spark/commit/293f1a4ebdd355152127e8089c2520ce0ff1bf15).

SparkQA commented:
**[Test build #56098 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/56098/consoleFull)** for PR 10910 at commit [`293f1a4`](https://github.com/apache/spark/commit/293f1a4ebdd355152127e8089c2520ce0ff1bf15).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/56098/
Test PASSed.

sarutak commented:
LGTM. If we have no other comments within 24 hours, I'll merge this.

sarutak commented:
Merging this into `master`. Thanks @ajbozarth !
